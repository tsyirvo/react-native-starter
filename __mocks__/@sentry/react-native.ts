module.exports = {
  ReactNavigationInstrumentation: jest.fn().mockImplementation(() => {}),
};
