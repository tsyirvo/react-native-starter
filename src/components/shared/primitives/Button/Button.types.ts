import {
  SpaceProps,
  LayoutProps,
  ColorProps,
  FlexboxProps,
  BorderProps,
  BorderRadiusProps,
  PositionProps,
  TypographyProps,
} from 'styled-system';

export type ButtonProps = SpaceProps &
LayoutProps &
ColorProps &
BorderProps &
BorderRadiusProps &
PositionProps &
TypographyProps &
FlexboxProps;

export type VariantProps = {
  variant?: 'base' | 'baseDisabled';
};
