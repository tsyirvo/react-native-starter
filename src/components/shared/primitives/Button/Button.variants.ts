import theme from '$styles/theme';

const baseStyles = {
  paddingHorizontal: theme.space.medium,
  paddingVertical: theme.space.small,
  borderRadius: theme.radii.medium,
};

const variants = {
  base: {
    backgroundColor: theme.colors.grey,
    ...baseStyles,
  },
  baseDisabled: {
    backgroundColor: theme.colors.grey,
    ...baseStyles,
  },
};

export default variants;
