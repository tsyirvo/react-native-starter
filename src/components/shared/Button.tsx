import React, { ReactElement } from 'react';
import { Pressable, PressableProps } from 'react-native';

type Props = PressableProps & {
  children: ReactElement;
  testID?: string | undefined;
};

const Button = ({ children, testID, ...rest }: Props) => (
  <Pressable testID={testID} {...rest}>
    {children}
  </Pressable>
);

Button.defaultProps = {
  testID: undefined,
};

export default Button;
