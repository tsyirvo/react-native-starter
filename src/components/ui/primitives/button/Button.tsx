/* eslint-disable react/jsx-props-no-spreading */

import {
  createRestyleComponent,
  createVariant,
  VariantProps,
} from '@shopify/restyle';
import { Pressable } from 'react-native';

import Box from '$components/ui/primitives/box/Box';
import Text from '$components/ui/primitives/text/Text';
import { Theme } from '$styles/theme';

import { BoxProps } from '../box/Box';

type ButtonProps = BoxProps &
  VariantProps<Theme, 'buttonVariants'> & {
    onPress: () => void;
    label: string;
    testID?: string;
    isEnabled?: boolean;
  };

const ButtonVariant = createVariant({
  themeKey: 'buttonVariants',
});

const PrimitiveButton = createRestyleComponent<
  VariantProps<Theme, 'buttonVariants'> & React.ComponentProps<typeof Box>,
  Theme
>([ButtonVariant]);

const Button = ({
  onPress,
  label,
  variant = 'base',
  testID,
  isEnabled = true,
  ...rest
}: ButtonProps) => (
  <PrimitiveButton variant={variant}>
    <Pressable disabled={!isEnabled} testID={testID} onPress={onPress}>
      <Box {...rest}>
        <Text>{label}</Text>
      </Box>
    </Pressable>
  </PrimitiveButton>
);

export default Button;
