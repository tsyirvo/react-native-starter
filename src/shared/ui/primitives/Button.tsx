/* eslint-disable react/jsx-props-no-spreading */

import type { VariantProps } from '@shopify/restyle';
import { createRestyleComponent, createVariant } from '@shopify/restyle';
import React from 'react';
import { Pressable } from 'react-native';

import type { Theme } from '$core/theme';

import type { BoxProps } from './Box';
import { Box } from './Box';
import { Text } from './Text';

interface ButtonProps extends BoxProps, VariantProps<Theme, 'buttonVariants'> {
  onPress: () => void;
  label: string;
  testID?: string;
  isEnabled?: boolean;
}

const ButtonVariant = createVariant({
  themeKey: 'buttonVariants',
});

const PrimitiveButton = createRestyleComponent<
  VariantProps<Theme, 'buttonVariants'> & React.ComponentProps<typeof Box>,
  Theme
>([ButtonVariant]);

export function Button({
  onPress,
  label,
  variant = 'base',
  testID,
  isEnabled = true,
  ...rest
}: ButtonProps) {
  return (
    <PrimitiveButton variant={variant}>
      <Pressable disabled={!isEnabled} testID={testID} onPress={onPress}>
        <Box {...rest}>
          <Text>{label}</Text>
        </Box>
      </Pressable>
    </PrimitiveButton>
  );
}
