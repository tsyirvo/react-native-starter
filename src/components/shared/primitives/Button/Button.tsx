/* eslint-disable react/jsx-props-no-spreading */

import {
  useRestyle,
  spacing,
  border,
  backgroundColor,
  SpacingProps,
  BorderProps,
  BackgroundColorProps,
  VariantProps,
  createVariant,
  createRestyleComponent,
  layout,
  LayoutProps,
  SpacingShorthandProps,
  spacingShorthand,
  BackgroundColorShorthandProps,
  backgroundColorShorthand,
} from '@shopify/restyle';
import { Pressable } from 'react-native';

import { Box, Text } from '$components/shared/primitives';
import { Theme } from '$styles/theme';

const restyleFunctions = [
  spacing,
  spacingShorthand,
  backgroundColor,
  backgroundColorShorthand,
  border,
  layout,
];

type BaseProps = SpacingProps<Theme> &
  SpacingShorthandProps<Theme> &
  BackgroundColorProps<Theme> &
  BackgroundColorShorthandProps<Theme> &
  BorderProps<Theme> &
  LayoutProps<Theme>;

type ButtonProps = BaseProps &
  VariantProps<Theme, 'buttonVariants'> & {
    onPress: () => void;
    label: string;
    testID?: string;
  };

const ButtonVariant = createVariant({
  themeKey: 'buttonVariants',
});

const PrimitiveButton = createRestyleComponent<
  VariantProps<Theme, 'buttonVariants'> & React.ComponentProps<typeof Box>,
  Theme
>([ButtonVariant]);

const Button = ({
  onPress,
  label,
  variant = 'base',
  testID,
  ...rest
}: ButtonProps) => {
  const props = useRestyle(restyleFunctions, rest);

  return (
    <PrimitiveButton variant={variant}>
      <Pressable testID={testID} onPress={onPress}>
        <Box {...props}>
          <Text>{label}</Text>
        </Box>
      </Pressable>
    </PrimitiveButton>
  );
};

export default Button;
