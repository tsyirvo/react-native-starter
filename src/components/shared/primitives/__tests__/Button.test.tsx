/* eslint-disable react/jsx-props-no-spreading */

import { fireEvent } from '@testing-library/react-native';
import React from 'react';
import { Text } from 'react-native';

import render from '$tests/utils';

import Button from '../Button';

describe('Shared primitives Button component', () => {
  // given
  const props = { onPress: jest.fn() };
  const child = <Text>Some text</Text>;

  it('should render correctly with a text children', () => {
    // When
    const wrapper = render(<Button {...props}>{child}</Button>);

    // Then
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger the onPress props when pressed', () => {
    // When
    const { getByText } = render(<Button {...props}>{child}</Button>);

    fireEvent.press(getByText('Some text'));

    // Then
    expect(props.onPress).toHaveBeenCalled();
  });
});
