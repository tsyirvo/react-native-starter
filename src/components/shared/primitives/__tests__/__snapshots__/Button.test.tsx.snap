// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shared primitives Button component should render correctly with a text children 1`] = `
<View
  accessible={true}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#e5e5e5",
        "borderRadius": 16,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      },
    ]
  }
  variant="base"
>
  <Text>
    Some text
  </Text>
</View>
`;
